homeassistant/components/neato/config_flow.py

```diff
@@ -1,112 +1,67 @@
-"""Config flow to configure Neato integration."""
-
+"""Config flow for Neato Botvac."""
 import logging
-
-from pybotvac import Account, Neato, Vorwerk
-from pybotvac.exceptions import NeatoLoginException, NeatoRobotException
-import voluptuous as vol
+from typing import Optional
 
 from homeassistant import config_entries
-from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers import config_entry_oauth2_flow
 
 # pylint: disable=unused-import
-from .const import CONF_VENDOR, NEATO_DOMAIN, VALID_VENDORS
-
-DOCS_URL = "https://www.home-assistant.io/integrations/neato"
-DEFAULT_VENDOR = "neato"
+from .const import NEATO_DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
 
-class NeatoConfigFlow(config_entries.ConfigFlow, domain=NEATO_DOMAIN):
-    """Neato integration config flow."""
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=NEATO_DOMAIN
+):
+    """Config flow to handle Neato Botvac OAuth2 authentication."""
 
-    VERSION = 1
+    VERSION = 2
+    DOMAIN = NEATO_DOMAIN
     CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
 
-    def __init__(self):
-        """Initialize flow."""
-        self._username = vol.UNDEFINED
-        self._password = vol.UNDEFINED
-        self._vendor = vol.UNDEFINED
+    _reauth = False
 
-    async def async_step_user(self, user_input=None):
-        """Handle a flow initialized by the user."""
-        errors = {}
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
 
-        if self._async_current_entries():
+    async def async_step_user(self, user_input: Optional[dict] = None) -> dict:
+        """Create an entry for the flow."""
+        current_entries = self._async_current_entries()
+        if current_entries and current_entries[0].version == self.VERSION:
+            # Already configured
             return self.async_abort(reason="already_configured")
 
-        if user_input is not None:
-            self._username = user_input["username"]
-            self._password = user_input["password"]
-            self._vendor = user_input["vendor"]
-
-            error = await self.hass.async_add_executor_job(
-                self.try_login, self._username, self._password, self._vendor
+        return await super().async_step_user(user_input=user_input)
+
+    async def async_step_reauth(self, config_entry: ConfigEntry) -> dict:
+        """Perform reauth upon migration of old entries."""
+        self._reauth = True
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: Optional[dict] = None
+    ) -> dict:
+        """Confirm reauth upon migration of old entries."""
+        if self._reauth:
+            self._reauth = False
```

Comment:
We don't need this. Check for user input is None instead.

Reply:
The `user_input` was always `None`

Reply:
The user input should be an empty dict when the user confirms the form.

Reply:
Unfortunately it isn't. It's always `None`.

Reply:
Then something is wrong.

Reply:
It's always `None`. This is how the form looks:

![grafik](https://user-images.githubusercontent.com/15657682/102267341-0b869100-3f1a-11eb-8191-7757fc7b37da.png)

There is no difference even with the addition in `strings.json` or with an empty dict as `data_schema`.

Reply:
That message is a different reason. That's shown when the integration dependencies are being installed before the flow is ready.

Reply:
I know. I just want to show how the whole page looks like. Tbh I have no idea how to create user_input.

Also, I think the current solution is not perfect, but better because it is not implicit. But if you disagree, I will do it as you say. You are the reviewer  :)